telemetry-proxy/telemetry_proxy.go

```diff
@@ -20,6 +22,117 @@ import (
 )
 
+type HistoricalDataRequest struct {
+	Start       int    `json:"start"` // Minutes ago
+	Stop        int    `json:"stop"`  // Minutes ago
+	Measurement string `json:"measurement"`
+	Field       string `json:"field"`
+	Aggregation string `json:"aggregation"` //e.g., "mean", "max", etc.
+	Every       int    `json:"every"`       // Seconds for aggregateWindow
+}
+
+type DataPoint struct {
+	Timestamp time.Time `json:"timestamp"`
+	Value     float64   `json:"value"`
+}
+
+type HistoricalDataResponse struct {
+	Data  []DataPoint `json:"data"`
+	Error string      `json:"error"`
+}
+
+type Datastore struct {
+	ctx          context.Context
+	token        string
+	influxUrl    string
+	influxClient influxdb2.Client
+	org          string
+	bucket       string
+	writeAPI     api.WriteAPIBlocking
+	queryAPI     api.QueryAPI
+}
+
+var datastore Datastore
+
+// Init initializes the struct with default values
+func (d *Datastore) Init(ctx context.Context) {
+	// Set up a connection to the influxdb instance.
+	d.ctx = ctx
+	d.token = os.Getenv("INFLUXDB_TOKEN")
+	d.influxUrl = "http://" + os.Getenv("INFLUXDB_HOSTNAME") + ":8086"
+	d.influxClient = influxdb2.NewClient(d.influxUrl, d.token)
+
+	d.org = "crt"
+	d.bucket = "telemetry"
+	d.writeAPI = d.influxClient.WriteAPIBlocking(d.org, d.bucket)
+	d.queryAPI = d.influxClient.QueryAPI(d.org)
+}
+
+// Store parses a packet and writes it to InfluxDB
+func (d *Datastore) Store(packet *pb.Telemetry) {
+	// Write to InfluxDB
+	tags := map[string]string{}
+	fields := map[string]interface{}{
+		"temp": packet.Temp,
+	}
+	point := write.NewPoint("temperature", tags, fields, time.Now())
+	writeCtx, writeCancel := context.WithTimeout(d.ctx, time.Second)
+	defer writeCancel()
+	if err := d.writeAPI.WritePoint(writeCtx, point); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// Query parses and executes a json request for historical data
+func (d *Datastore) Query(req HistoricalDataRequest) HistoricalDataResponse {
+	// Validate input (important to prevent injection attacks!)
+	if req.Measurement == "" || req.Field == "" {
+		response := HistoricalDataResponse{Error: "Measurement and field are required"}
+		return response
+	}
+
+	if req.Aggregation == "" {
+		req.Aggregation = "mean" // Default to mean if not provided
+	}
+
+	if req.Every == 0 {
+		// Default to 100 data points
+		numDatapoints := 100.0
+		if width := int((float64(req.Stop-req.Start) / numDatapoints) * 60); width != 0 {
+			req.Every = width
+		} else {
+			req.Every = 1
+		}
+	}
+
+	// Construct query
+	query := fmt.Sprintf(`from(bucket: "telemetry")
+	  |> range(start: %dm, stop: %dm)
+	  |> filter(fn: (r) => r._measurement == "%s")
+	  |> filter(fn: (r) => r._field == "%s")
+	  |> aggregateWindow(every: %ds, fn: %s, createEmpty: false)
+	  |> drop(columns: ["table", "_measurement", "_field", "_start", "_stop"])
+	  |> yield(name: "mean")`, req.Start, req.Stop, req.Measurement, req.Field, req.Every, req.Aggregation)
```

Comment:
Also my worry with raw string format querying like this is that it is easy to escape. For example, if I make `req.Start` equal to:

```
5m, stop: 2m)
|> join(...)
```

(Assuming there is a join command or something similar) I can now get any other datapoint in the database. Granted it is not that important now that we only have historical data. But maybe there is a delete command, or if we store users in database later this is not good.

Reply:
Agreed on users being more strict. Shouldn't be any editing issues since the QueryAPI is separate from the WriteAPI. I think we make a ticket and tackle this later. Some ideas for handling this are checking to make sure there are no pipes (|>) or taking ints for measurement and field and indexing in to an array. Your example with start shouldn't be a problem since we marshal to an int, so it ignores string inputs there. Even with the string fields, I can't manage to escape it. The marshaler seems good at rejecting those attempts